sopel/irc/isupport.py

```diff
@@ -0,0 +1,380 @@
+# coding=utf-8
+"""IRC Tools for ISUPPORT management.
+
+When a server wants to advertise its features and settings, it can use the
+``RPL_ISUPPORT`` command with a list of arguments.
+
+.. seealso::
+
+    https://modern.ircdocs.horse/#rplisupport-005
+
+"""
+# Copyright 2019, Florian Strzelecki <florian.strzelecki@gmail.com>
+#
+# Licensed under the Eiffel Forum License 2.
+from __future__ import unicode_literals, absolute_import, print_function, division
+
+import functools
+import itertools
+import re
+
+
+def _optional(parser, default=None):
+    # set a parser as optional: will always return the default value provided
+    # if there is no value (empty or None)
+    @functools.wraps(parser)
+    def wrapped(value):
+        if not value:
+            return default
+        return parser(value)
+    return wrapped
+
+
+def _no_value(value):
+    # always ignore the value
+    return None
+
+
+def _single_character(value):
+    if len(value) > 1:
+        raise ValueError('Too many characters: %r.' % value)
+
+    return value
+
+
+def _map_items(parser=str, map_separator=',', item_separator=':'):
+    @functools.wraps(parser)
+    def wrapped(value):
+        items = sorted(
+            item.split(item_separator)
+            for item in value.split(map_separator))
+
+        return tuple(
+            (k, parser(v) if v else None)
+            for k, v in items
+        )
+    return wrapped
+
+
+def _parse_chanmodes(value):
+    items = value.split(',')
+
+    if len(items) < 4:
+        raise ValueError('Not enough channel types to unpack from %r.' % value)
+
+    # add extra channel type's modes to their own tuple
```

Comment:
```suggestion
    # add extra channel mode types to their own tuple
```

Reply:
As I was reading the doc, they talk about "channel type" having "modes". I'm not sure `channel mode types` convey the same meaning. How can I be sure?

Reply:
I'm not sure exactly what documentation you used for reference, but the four standard categories you get from the CHANMODES parameter _are_ different types of _mode_:

<ol type="A">
  <li>Modes that add or remove an address from a list (MUST have a parameter unless the client is sending them without one to obtain the current list)</li>
  <li>Modes that change a channel setting (MUST have a parameter, always)</li>
  <li>Modes that change a channel setting (MUST be set with a parameter, and MUST NOT have a parameter when unset</li>
  <li>Modes that change a channel setting (MUST NOT have a parameter)</li>
</ol>

Neither [the Internet draft](https://tools.ietf.org/html/draft-hardy-irc-isupport-00#section-4.3) I checked (linked from my go-to [list of numerics](https://www.alien.net.au/irc/irc2numerics.html), which always comes up in my searches for "irc numerics") nor [modern.ircdocs.horse](https://modern.ircdocs.horse/#chanmodes-parameter) says anything about "channel type" in their respective sections about the CHANMODES parameter. They are categories/types of _mode_ only.

Reply:
Ooooh okay now I get it. My mistake!